### DISCUSSION TITLE
Worker. WASM C code ending callback

### DISCUSSION THREAD

**Discussion by DiachenkoIgor** [emscripten-core/emscripten#17862] [Q&A]
Hi, when I create thread in C, new worker is created. JS script with name *.worker.js call function 'Module['invokeEntryPoint']' which start C procedure for given thread. However, sometimes this function is blocking and releases after C code execution end , another time function is released without ending C code. Can some, please, explain how to predict behaviour of this function or advice other way of defining end of execution of C code in worker?

**Comment by sbc100** [comment]
Can you give some more details.  What does "function is released without ending C code" mean?   Can you give an example?   Can you share the link flags you are using?

**Reply by DiachenkoIgor** [reply]
Linker flags: -sPROXY_TO_PTHREAD -sPTHREAD_POOL_SIZE=1 -sUSE_PTHREADS=1 -pthread

Code example:
`void *print_message_function( void *ptr )
{
     for(int i = 0; i < 10; i++){
       printf("Sleeping...\n");
     sleep(5);
     }
     return NULL;
}`

"invokeEntryPoint" return value before 'printf' and 'return NULL' execution.

**Reply by sbc100** [reply]
Can you show the full example?   What does your main function do?  

I assume `invokeEntryPoint` is being called during normal pthread creation, and you not calling it yourself somehow?

**Reply by DiachenkoIgor** [reply]
You are right.

Full example:

```
void *print_message_function( void *ptr );

int main()
{
     pthread_t thread1;

     pthread_create( &thread1, NULL, print_message_function, (void*) NULL);

     pthread_join(thread1, NULL);

     printf("END...\n");

     return 0;
}

void *print_message_function( void *ptr )
{
     for(int i = 0; i < 10; i++){
     printf("Sleeping...\n");
     sleep(5);
     }
     return NULL;
}
```

**Reply by sbc100** [reply]
That example seems to work fine for me:

```
sbc.sbc1 (~/dev/wasm/emscripten) (wasm2js_info) $ cat test.c 
#include <pthread.h>
#include <unistd.h>
#include <stdio.h>

void *print_message_function( void *ptr );

int main()
{
   pthread_t thread1;

   pthread_create( &thread1, NULL, print_message_function, (void*) NULL);

   pthread_join(thread1, NULL);

   printf("END...\n");

   return 0;
}

void *print_message_function( void *ptr )
{
   for (int i = 0; i < 3; i++){
     printf("Sleeping...\n");
     sleep(5);
   }
   return NULL;
}
$ emcc  -sPROXY_TO_PTHREAD -sPTHREAD_POOL_SIZE=1 -sUSE_PTHREADS=1 -pthread test.c
$ node --experimental-wasm-threads --experimental-wasm-bulk-memory ./a.out.js 
Sleeping...
Sleeping...
Sleeping...
END...
Pthread 0x501368 completed its main entry point with an `unwind`, keeping the worker alive for asynchronous operation.
program exited (with status: 0), but EXIT_RUNTIME is not set, so halting execution but not exiting the runtime or preventing further async execution (build with EXIT_RUNTIME=1, if you want a true shutdown)
```

**Reply by DiachenkoIgor** [reply]
Sorry for waiting. I don't know why such situation occurs. Can you please advice about way to be notified of ending execution of C code in worker, because I need custom resource handling of js environment?

**Reply by sbc100** [reply]
pthread_join should not return until the thread is complete.   The example you posted seems to work fine at least when I tested it as above?   Are doing anything else unusual?  perhaps you should share the full failing example, in a github repo is you like?

**Reply by DiachenkoIgor** [reply]
pthread_join works fine. worker js 'invokeEntryPoint' function not works fine. Unfortunately, my project is very big and hard, so I will make more research and will reopen discussion, ok ?

**Reply by sbc100** [reply]
Are you calling `invokeEntryPoint` yourself?   Can you explain how/why you are doing that?  Its really only designed to be called in worker.js as an implementation detail of pthread_create.

**Reply by DiachenkoIgor** [reply]
Hi, I reproduced such behaviour.
test.cpp

```
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>
extern "C" int test_asyncify();
void *print_message_function( void *ptr );

int main()
{
     pthread_t thread1;

     pthread_create( &thread1, NULL, print_message_function, (void*) NULL);

     pthread_join(thread1, NULL);

     printf("END...\n");

     return 0;
}

void *print_message_function( void *ptr )
{
     for(int i = 0; i < 5; i++){
          printf("print_message...\n");
          sleep(1);
     }
     test_asyncify();

     printf("After test_asyncify...\n");

     return NULL;
}
```
test_library.js
```
mergeInto(LibraryManager.library, {

test_asyncify: function () {
 return Asyncify.handleAsync(async ()=>{await new Promise(resolve => {
    setTimeout(() => {
      resolve('resolved');
    }, 5000);
  })});
}
});
```
compile command
`emcc test.cpp -s PROXY_TO_PTHREAD -s ASYNCIFY -s ASYNCIFY_IMPORTS=test_asyncify --js-library ./test_library.js -sPTHREAD_POOL_SIZE=1 -sUSE_PTHREADS=1 -pthread -o test.html`

pthread_join return before thread method end execution.

**Reply by sbc100** [reply]
Ah yes, this looks like an issue with ASYNCIFY.   The way ASYNCIFY works is that it re-writes the wasm file to return early on each suspend and then ti re-winds to where it left off after the suspend.    I'm not sure ASYNCIFY works with pthreads today,   If this is true we should make this more clear.  @kripken do you know if ASYNCIFY is supposed to work with pthreads today?

**Reply by kripken** [reply]
I don't believe we've tested asyncify+pthreads so far. Very possible things are not working yet, though I can't think offhand of a particular issue relevant here.

**Reply by DiachenkoIgor** [reply]
Is there other way for implementation such tasks?

**Reply by mmarczell-graphisoft** [reply]
I think we've tried asyncify + pthreads at my company and the only thing that prevented us from using it in production was https://github.com/emscripten-core/emscripten/issues/16291 and that was since fixed.
